### PR TITLE
api: /v1/user/initialize to create first user (PROJQUAY-1926) 

### PR DESCRIPTION
**Pull request by thomasmckay**

### CODE REVIEW

File: endpoints/api/user.py [hunk 1]
```diff
@@ -545,6 +546,78 @@ def delete(self):
         return "", 204
 
 
+@resource("/v1/user/initializexxxx")
```
Comment (jonathankingfc) [inline endpoints/api/user.py:549]: do we still need this endpoint? or is the endpoint defined in `web.py` sufficient?

File: endpoints/web.py [hunk 2]
```diff
@@ -897,3 +898,60 @@ def redirect_to_namespace(namespace):
...
+    Create initial user in an empty database
+    """
+
+    # Ensure that we are using database auth.
+    if app.config["AUTHENTICATION_TYPE"] != "Database":
+        response = jsonify({"message": "Cannot initialize user in a non-database auth system"})
+        response.status_code = 400
+        return response
+
+    # if model.user.get_active_user_count() != 0:
```
Comment (jonathankingfc) [inline endpoints/web.py:915]: is this intentionally commented out? can we use the same function as in the `user.py` endpoint defined above?

**Review by jonathankingfc — CHANGES_REQUESTED** [review]

File: endpoints/web.py (hunk 2 above)
Comment (thomasmckay) [reply endpoints/web.py:915]: intentional yes during my testing so i don't have to reset db constantly :)

File: endpoints/api/user.py (hunk 1 above)
Comment (thomasmckay) [reply endpoints/api/user.py:549]: still trying to figure out how this method can be included in user.py instead of web.py. one or the other will "win"

File: endpoints/api/user.py [hunk 3]
```diff
@@ -545,6 +546,83 @@ def delete(self):
...
+            success, headers = common_login(new_user.uuid)
+            if not success:
+                raise request_error(message="Could not login. Failed to initialize user")
+
+            return (
+                {
+                    "username": user_data["username"],
+                    "email": user_data.get("email"),
+                    "password": user_data["password"],
+                    "encrypted_password": authentication.encrypt_user_password(
```
Comment (kleesc) [inline endpoints/api/user.py:615]: **Don't return passwords**

File: endpoints/api/user.py [hunk 4]
```diff
@@ -545,6 +546,83 @@ def delete(self):
...
+    def post(self):
+        """
+        Create initial user in an empty database
+        """
+
+        # Ensure that we are using database auth.
+        if app.config["AUTHENTICATION_TYPE"] != "Database":
+            raise request_error(message="Cannot initialize user in a non-database auth system")
+
+        if model.user.get_active_user_count() != 0:
```
Comment (kleesc) [inline endpoints/api/user.py:591]: This is doing a full table count every time someone hits this endpoint. Especially, since anyone can hit this api, nothing prevents from constantly hitting this and exhausting db resources.

Since you don't really care about the actual count, something like `User.select().limit(1)` would be more efficient.

File: endpoints/api/user.py [hunk 5]
```diff
@@ -57,6 +57,7 @@
     define_json_response,
     RepositoryParamResource,
     page_support,
+    verify_not_prod,
```
Comment (kleesc) [inline endpoints/api/user.py:60]: Let's not have specific code for production vs enterprise installs

File: endpoints/api/__init__.py [hunk 6]
```diff
@@ -64,7 +64,7 @@ def handle_error(self, error):
 api = ApiExceptionHandlingApi()
 api.init_app(api_bp)
 api.decorators = [
-    csrf_protect(),
+    csrf_protect(skip_paths=["/api/v1/user/initialize"]),
```
Comment (kleesc) [inline endpoints/api/__init__.py:67]: Instead of hacking in the current decorator for one specific use case, if your api needs a different set of decorators than these ones, you should create a separate blueprint for your resource.

Comment (kleesc) [reply endpoints/api/__init__.py:67]: Also, for this case, you can probably make the request on behalf of Quay, and use  `app.config["HTTPCLIENT"]` to go around the CRSF protection.

File: endpoints/api/user.py [hunk 7]
```diff
@@ -545,6 +546,83 @@ def delete(self):
         return "", 204
 
 
+@resource("/v1/user/initialize")
+class InitializeUser(ApiResource):
```
Comment (kleesc) [inline endpoints/api/user.py:550]: This api should be put behind a feature flag, and defaults to `False`

**Review by kleesc — CHANGES_REQUESTED** [review]

File: endpoints/test/test_csrf.py [hunk 8]
```diff
@@ -1,8 +1,66 @@
...
+        # POST calls are checked
+        (
+            "/api/something/else",
+            "POST",
+            ["/api/something"],
+            False,
+        ),
+    ],
+)
+def test_csrf_protect(path, method, skip_paths, success):
```
Comment (kleesc) [inline endpoints/test/test_csrf.py:53]: See comment about Flask blueprints

**Review by kleesc — CHANGES_REQUESTED** [review]

File: endpoints/api/__init__.py (hunk 6 above)
Comment (thomasmckay) [reply endpoints/api/__init__.py:67]: Do you have example of HTTPCLIENT and how it would be used?

File: endpoints/api/user.py (hunk 3 above)
Comment (thomasmckay) [reply endpoints/api/user.py:615]: I'll need the `encrypted_password` returned for use in follow-up API calls (that's the bearer token, right?). I can remove the `password`.

File: endpoints/api/user.py (hunk 7 above)
Comment (thomasmckay) [reply endpoints/api/user.py:550]: For what purpose? It only functions on a clean db.

**Comment by kleesc** [comment]
Not sure how being able to create the first user from api will help. You'll still need to go through the auth flow through the ui to be able to create an org/oauth token anyways...

File: endpoints/api/user.py (hunk 3 above)
Comment (kleesc) [reply endpoints/api/user.py:615]: No. Passwords and tokens are different.

File: endpoints/csrf.py [hunk 9]
```diff
@@ -60,13 +60,19 @@ def csrf_protect(
     request_token_name=_QUAY_CSRF_TOKEN_NAME,
     all_methods=False,
     check_header=True,
+    skip_paths=None,
```
Comment (jonathankingfc) [inline endpoints/csrf.py:63]: Do we still need this file?

Comment (thomasmckay) [reply endpoints/csrf.py]: fixed, thanks

**Review by jonathankingfc — DISMISSED** [review]
LGTM

File: endpoints/api/test/test_user.py [hunk 10]
```diff
@@ -40,3 +41,106 @@ def test_user_metadata_update(client):
...
+                    "web.user_initialize",
+                    url_for,
+                    "POST",
+                    {},
+                    body=metadata,
+                    expected_code=expected_code,
+                    headers={"Content-Type": "application/json"},
+                )
+
+                if expected_code == 200:
```
Comment (kleesc) [inline endpoints/api/test/test_user.py:138]: An another assertion that the status code matches the expected one.

File: endpoints/web.py [hunk 11]
```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
...
+            model.oauth.create_user_access_token(
+                new_user, user_data["username"], scope, access_token=token_string
+            )
+            result["access_token"] = token_string
+
+        return (result, 200, headers)
+    except model.user.DataModelException as ex:
+        response = jsonify({"message": "Failed to initialize user: " + str(ex)})
+        response.status_code = 400
+        return response
```
Comment (kleesc) [inline endpoints/web.py:979]: Wasn't this previously in the `user.py` or `superuser.py` module? Why did you move it to `web`?

File: endpoints/api/test/test_user.py [hunk 12]
```diff
@@ -40,3 +41,106 @@ def test_user_metadata_update(client):
...
+                "password": "password",
+            },
+        ),
+        # Empty database with access token
+        (
+            0,
+            200,
+            False,
+            True,
+            {"username": "someone", "password": "password", "access_token": "true"},
```
Comment (kleesc) [inline endpoints/api/test/test_user.py:113]: Not sure what case this is. Shouldn't be able to have a token without a user..

File: endpoints/web.py (hunk 11 above)
Comment (thomasmckay) [reply endpoints/web.py:979]: This in web.py due to the csrf and api needing to be unauthenticated. We spent a long time back and forth on this.

File: endpoints/api/test/test_user.py (hunk 12 above)
Comment (thomasmckay) [reply endpoints/api/test/test_user.py:113]: This is testing success w/o mailing feature. Not sure what comment means.

File: endpoints/api/test/test_user.py (hunk 10 above)
Comment (thomasmckay) [reply endpoints/api/test/test_user.py:138]: this assert is done in `conduct_call()`

File: endpoints/web.py [hunk 13]
```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
...
+        if user_data.get("access_token"):
+            model.oauth.create_application(
+                new_user,
+                "automation",
+                "",
+                "",
+                client_id=user_data["username"],
+                description="Application token generated via /api/v1/user/initialize",
+            )
+            token_string = random_string_generator(length=40)()
```
[the remark this answers is not in the document]
Comment (thomasmckay) [reply endpoints/web.py:968]: done

File: endpoints/web.py [hunk 14]
```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
...
+            user_data.get("email"),
+            auto_verify=True,
+            email_required=features.MAILING,
+            is_possible_abuser=False,
+            prompts=prompts,
+        )
+        success, headers = common_login(new_user.uuid)
+        if not success:
+            response = jsonify({"message": "Could not login. Failed to initialize user"})
+            response.status_code = 400
```
[the remark this answers is not in the document]
Comment (thomasmckay) [reply endpoints/web.py:948]: done

File: endpoints/web.py [hunk 15]
```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
...
+
+
+def has_users():
+    """
+    Return false if no users in database yet
+    """
+    return bool(list(User.select().limit(1)))
+
+
+@web.route("/api/v1/user/initialize", methods=["POST"])
```
[the remark this answers is not in the document]
Comment (thomasmckay) [reply endpoints/web.py:910]: done

File: endpoints/web.py [hunk 16]
```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
         return redirect(url_for("web.org_view", path=namespace))
     else:
         return redirect(url_for("web.user_view", path=namespace))
+
+
+def has_users():
+    """
+    Return false if no users in database yet
+    """
+    return bool(list(User.select().limit(1)))
```
[the remark this answers is not in the document]
Comment (thomasmckay) [reply endpoints/web.py:907]: done

File: endpoints/api/test/test_user.py [hunk 17]
```diff
@@ -40,3 +41,94 @@ def test_user_metadata_update(client):
...
+                )
+
+                if expected_code == 200:
+                    assert user.json["username"] == metadata["username"]
+                    if feature_mailing:
+                        assert user.json["email"] == metadata["email"]
+                    else:
+                        assert user.json["email"] is None
+                    assert user.json.get("encrypted_password", None)
+                    if metadata.get("access_token"):
```
Comment (kleesc) [inline endpoints/api/test/test_user.py:133]: Can you add an assertion that checks the opposite is also true: if access_token is not set, then there shouldn't be one in the response.

Comment (thomasmckay) [reply endpoints/api/test/test_user.py:133]: done

**Review by kleesc — APPROVED** [review]
